lib/modules/versioning/pvp/index.ts

```diff
@@ -0,0 +1,297 @@
+import { logger } from '../../../logger';
+import type { RangeStrategy } from '../../../types/versioning';
+import { regEx } from '../../../util/regex';
+import type { NewValueConfig, VersioningApi } from '../types';
+
+export const id = 'pvp';
+export const displayName = 'Package Versioning Policy (Haskell)';
+export const urls = [];
+export const supportsRanges = true;
+export const supportedRangeStrategies: RangeStrategy[] = ['auto'];
+
+type Parsed = { lower: string; upper: string };
+type Components = { major: number[]; minor: number[]; patch: number[] };
+
+export function parse(input: string): Parsed | null {
+  const noSpaces = input.replaceAll(' ', '');
+  const r = regEx(/>=(?<lower>[\d.]+)&&<(?<upper>[\d.]+)/);
+  const m = r.exec(noSpaces);
+  if (!m?.groups) {
+    return null;
+  }
+  return {
+    lower: m.groups['lower'],
+    upper: m.groups['upper'],
+  };
+}
+
+export function extractAllComponents(version: string): number[] {
+  const versionMajor = version.split('.');
+  const versionIntMajor: number[] = versionMajor.map((x) => parseInt(x, 10));
+  const ret = [];
+  for (const l of versionIntMajor) {
+    if (l < 0 || !isFinite(l)) {
+      continue;
+    }
+    ret.push(l);
+  }
+  return ret;
+}
+
+function compareIntArray(
+  versionIntMajor: number[],
+  otherIntMajor: number[],
+): 'lt' | 'eq' | 'gt' {
+  for (
+    let i = 0;
+    i < Math.min(versionIntMajor.length, otherIntMajor.length);
+    i++
+  ) {
+    if (versionIntMajor[i] > otherIntMajor[i]) {
+      return 'gt';
+    }
+    if (versionIntMajor[i] < otherIntMajor[i]) {
+      return 'lt';
+    }
+  }
+  if (versionIntMajor.length === otherIntMajor.length) {
+    return 'eq';
+  }
+  if (versionIntMajor.length > otherIntMajor.length) {
+    return 'gt';
+  }
+  return 'lt';
+}
+
+function isGreaterThan(version: string, other: string): boolean {
+  const versionIntMajor = extractAllComponents(version);
+  const otherIntMajor = extractAllComponents(other);
+  return compareIntArray(versionIntMajor, otherIntMajor) === 'gt';
+}
+
+function getMajor(version: string): number {
+  // This basically can't be implemented correctly, since
+  // 1.1 and 1.10 become equal when converted to float.
+  // Consumers should use isSame instead.
+  const l1 = version.split('.');
+  return Number(l1.slice(0, 2).join('.'));
+}
+
+function getMinor(version: string): number | null {
+  const l1 = version.split('.');
+  if (l1.length < 3) {
+    return null;
+  }
+  return Number(l1[2]);
```

Comment:
Centralize the version parsing in a function that function should return an interface which holds that information. 
This can be reused for parsing the range elements.

Reply:
There is an existing function called `getComponents` that provides what you describe. I've pushed a commit such that this function is used.